File: 20210523LangReview/Python/06-closure/closure.swift
/*

a simple module demonstrates closure mechnsim in Swift

@ZL, 20210525

[ experiment result ]
~ by default, Swift allows inner scope to read the object in outer scope
~ Swift does allow re-binding or write an object in outer scope

*/

import cocoa
import UIKit

func outer(_ n: Int) -> () -> Int {
    var total: Int = 0
    func inner() -> Int {
        total += n
        return total
    }
    return inner
}

let f = outer(10)

print("1st call, \(f())")
print("2nd call, \(f())")
print("3rd call, \(f())")

File: 20210523LangReview/Python/06-closure/closure.py
"""
a simple module demonstrates closure mechnsim in Python

@ZL, 20210525

[ experiment result ]
^ Python prevents this behavior (access local variable) by default
~ by default, Python allows inner scope to read the object in outer scope
~ but Python does not allow re-binding or write an object in outer scope

& To achieve the same behavior
using keyword `global` or `nonlocal` to re-bind the object in outer scope 

"""

def func(n: int):
    total = 0
    def inner():
        nonlocal total
        total += n
        return total
    return inner

if __name__ == "__main__":
    myfunc = func(10)
    print(myfunc())
    print(myfunc())
    print(myfunc())